backends/ebpf/runtime/pcap_util.c

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2018 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stdlib.h>     // EXIT_SUCCESS, EXIT_FAILURE
+#include <string.h>     // memcpy()
+#include "pcap_util.h"
+
+#define DLT_EN10MB 1
+
+/* Dynamically-allocated list of packets.
+ */
+struct pcap_list {
+    pcap_pkt **pkts;
+    uint32_t len;
+};
+
+/* An array of lists of packets */
+struct pcap_list_array {
+    pcap_list_t **lists;
+    uint16_t len;
+};
+
+pcap_list_t *append_packet(pcap_list_t *pkt_list, pcap_pkt *pkt) {
+    if (!pkt_list)
+        /* If the list is not allocated yet, create it */
+        pkt_list = allocate_pkt_list();
+    pkt_list->len++;
+    pkt_list->pkts = realloc(pkt_list->pkts, pkt_list->len * sizeof(pcap_pkt *));
+    if (pkt_list->pkts == NULL) {
+        fprintf(stderr, "Fatal: Failed to expand the"
+            "packet list with size %u !\n", pkt_list->len);
+        exit(EXIT_FAILURE);
+    }
+    pkt_list->pkts[pkt_list->len - 1] = pkt;
+    return pkt_list;
+}
+
+pcap_list_array_t *insert_list(pcap_list_array_t *pkt_array, pcap_list_t *pkt_list, uint16_t index) {
+    if (!pkt_array)
+        /* If the array is not allocated yet, create it */
+        pkt_array = allocate_pkt_list_array();
+    if (index >= pkt_array->len) {
+        pkt_array->lists = realloc(pkt_array->lists, (index+1) * sizeof(pcap_list_t *));
+        if (pkt_array->lists == NULL) {
```

Comment:
You should set all lists between the current size and index to null. Currently there is garbage there.

Reply:
I decided against that since I set the value with  pkt_list->pkts[pkt_list->len - 1] = pkt in the next line anway. So it seemed superfluous

Reply:
What if you add a packet to list number 10 and lists 0 to 9 do not exist?